[JUJU-1889] Log forwarding, take two

barrettj12 opened the pull request:
This PR enables forwarding of Pebble service logs to remote Loki servers and UDP/TCP syslog receivers, following the spec [here](https://docs.google.com/document/d/12tJFhIbVIlOcblG6l9_u-K5bCib78lojRJ-hdGBxSsw/edit?usp=sharing).

## Design

Each (service, log target) pair has an associated "forwarder", which pulls logs from the service ringbuffer (using the iterators introduced in #29), and sends them to the log target. Each forwarder runs two goroutines:
- an `iterate` goroutine, which pulls logs from the iterator/parser and sends them on a `messages` channel
- a `forward` goroutine, which reads logs from the `messages` channel, buffers them, and flushes periodically to the remote log target (after 100 logs have been buffered, or 1 second after the first write)

The forwarders are controlled by a "log manager", which is notified when a new service starts, or when the plan changes. On a plan change, the log manager tears down all forwarders, and starts new ones as determined by the plan.

## Plan changes

The Pebble layer specification has changed as follows:
- A new top-level section `log-targets`, where users specify the log targets they want Pebble to interact with.
- A new (optional) field `log-targets` in the service spec. This takes a list of names, which are matched against the targets defined in the `log-targets` top-level section.

The default behaviour is to forward all services' logs to all targets. This makes it easy to get started with log forwarding by just defining a new layer with the log targets - no need to change the service definitions.

Users can have more granular control over log forwarding by:
- Setting the optional `selection` field for a log target;
- In the definition for a service, explicitly listing log targets they want to forward this service's logs to.

More details are in the README.

## Other changes

The `plan.Service.Equal` method was being used by the service manager to determine whether or not to restart a service on a replan. However, not all changes to the service config require a service restart. In particular, a change to the new `log-targets` field should **not** invoke a service restart. Hence, we've defined a `servstate.NeedsRestart(old, new *plan.Service)` function, which compares the old and new config structs for a service and determines if the service needs to be restarted. At the moment, `NeedsRestart` only ignores changes to `log-targets` - future work should also ignore changes to e.g. metadata, so we can be more conservative with restarting services.

This was the only use of `plan.Service.Equal`, hence we've also removed that method.

## Future work

- Add support for custom Loki labels ([spec](https://docs.google.com/document/d/1HUHH9OQW-czPxn1mXr6AVjhwRMM-DRYp02ATllFSxjE/edit)).
- Allow configuring the 100 log limit / 1 second timeout in the plan (?).

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+//   - we have maxMessages ready to send; or
+//   - delay has passed since the last request.
+type lokiClient struct {
+	lokiURL     string
```
I think we should make this a hard URL type.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+	maxMessages int
+	delay       time.Duration
+
+	// buffered request so far
```
All of these members are going to need concurrency locks.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+		maxMessages: maxMessages,
+		delay:       delay,
+	}
+	c.Start()
```
Probably not the greatest idea to start a go routine in a new method. Should probably be made an explicit call by the creator.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func (c *lokiClient) reset() {
+	c.buffer = bytes.NewBuffer([]byte(requestStart))
```
Needs locking

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+	go c.run()
+}
+
+func (c *lokiClient) reset() {
```
Reset should only work when there is no go routine running. This one is going to be tricky to deal with I think.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+func (c *lokiClient) run() {
+	for {
+		select {
+		case <-c.writeCh:
```
Nit:

```
case <-c.writeCh:
case <-c.timeCh:
    c.flush()
```

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+
+	c.buffer.WriteString(requestEnd)
+	logger.Noticef(c.buffer.String())
+	resp, err := http.Post(c.lokiURL, "application/json", c.buffer)
```
We shouldn't use the default go http client. One should be made with the lokiClient.

barrettj12 replied:
Why, what's wrong with the default client?
@hpidcock thoughts?

benhoyt replied:
It may or may not matter here, but I recommend not using `http.DefaultClient`, because you don't get control over timeouts (unless you set the global timeout) and connection pooling, and it's harder to test because you can't pass in a testing client. Similar to the problems with globals in general.

You can work around the timeout issue by using `http.NewRequestWithContext` and supplying a context without a timeout, though. Still, I'd advise against it, and prefer passing in the `*http.Client` "from above" as a dependency.

barrettj12 replied:
ok sure, easy enough to use a custom client

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+	}
+
+	c.buffer.WriteString(requestEnd)
+	logger.Noticef(c.buffer.String())
```
Does this create a feed back loop by making logs?

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+		logger.Noticef("error sending HTTP request: %v", err)
+	}
+	logger.Noticef("response from Loki: %v", resp.Status)
+	if resp.StatusCode >= 400 {
```
Nit:

Use the status constants defined in the http package.

barrettj12 replied:
not so sure about that, those constants are for specific HTTP codes not ranges

barrettj12 replied:
`>= 400` indicates error - 4xx client error, 5xx server error. The idea is to get more info if we get an error response

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+	}
+	logger.Noticef("response from Loki: %v", resp.Status)
+	if resp.StatusCode >= 400 {
+		b, _ := io.ReadAll(resp.Body)
```
Need to handle the error.

barrettj12 replied:
this is just trying to get error info, if we can't read the body, then just give up IMO

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+		b, _ := io.ReadAll(resp.Body)
+		logger.Noticef("%s", b)
+	}
+	c.reset()
```
I think we are in a pretty bad place here when we add concurrency locking. You are going to end up in a dead lock situation I think.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+	if c.messages > 0 {
+		c.buffer.WriteString(",")
+	}
+	c.buffer.WriteString(`["`)
```
You can use Fprintf here to clean this up a lot and make it more readable.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,136 @@
+}
+
+func (c *lokiClient) WriteEntry(entry servicelog.Entry) {
+	// Write in format
```
Without concurrency controls here we are going to overwrite and corrupt memory being used in the send and also possible loose future logs due to the reset call from send.

tlm commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+func NewLogManager() *LogManager {
+	lm := &LogManager{
```
Nit:

`return &LogManager{}`

tlm commented:
Something to possibly consider from the conversations yesterday is my chaining concept. Looking at what we have here it's very obvious that a lot of the concurrency problems can be solved by implementing it. The design above has to introduce a mammoth amount of locking and magic to work safely. But with chaining we can get rid of all the locks.

If we start with out our log client interface and rejig it a bit to be.

```
type LogClient interface {
    Flush()
    WriteEntry(...servicelog.Enty)
}
```

We can then implement something like this for the buffer:
```
type LogClientBuffer struct {
    buffer []servicelog.Entry
    maxEntries uint
    next LogClient // This would be a reference to the loki client or another thing in the chain
}

func (l *LogClientBuffer) Flush() {
    // call next.WriteEntry with the contents of buffer
    // reset buffer
    // call flush on next
}

func (l *LogClientBuffer) WriteEntry(e ...servicelog.Entry) {
    // loop over all of e and append to buffer
   // if buffer is greater then maxEntries call flush
}
```

We can then use the pusher go routine to send all Write and Flush events to the client. It gives a single go routine to work off and also allows us to make clients that don't have to worry about being thread safe.

For the timeout of the buffer is we either A) put a cancel context on the iterator that calls flush on the client or B) we add a timeout to the pusher that calls Flush() every x seconds or C) we keep reading the next iterator and calling write till next blocks and then we call flush on the client.

We can chat about that last paragraph as a group to flesh out the best of those options. But I think doing it this way means the loki client doesn't need to implement buffering and we can get rid of the concurrency problems.

tlm commented on internal/overlord/logstate/pumper.go:
```diff
@@ -0,0 +1,115 @@
+		return nil, err
+	}
+
+	stopCh := make(chan os.Signal, 1)
```
There are a number of reasons why we shouldn't be creating a stop signal on os sig's.

The best one for explanation is this pumper has a very good chance of stopping before the service and so we will potentially drop log message from the service as it shuts down.

Stopping of a pumper should be tied into the normal shut down of pebble.

barrettj12 replied:
the whole point of this was to flush the buffer when Ctrl-C is pressed rather than dropping the messages

benhoyt replied:
If I understand what Tom is saying here, I agree. We should only handle OS signals in main at the very top level. Here we should be passing in a context or stop channel, and cancel/close that (in main, or passed down from main) when Ctrl-C is pressed.

barrettj12 replied:
I'm thinking each service should notify the log manager when it stops, then the log manager can stop the relevant pumpers. Can't quite work out where the actual service stopping code is though

tlm commented on internal/overlord/logstate/pumper.go:
```diff
@@ -0,0 +1,115 @@
+		}
+
+		// Let client flush if it's ready
+		p.client.MaybeFlush()
```
I am not a fan of MaybeFlush() wording or what it's trying to do here. Specifically it's a function that is going to get called at ad nauseam for no reason and depending on the log back end it could be a very expensive call.

I think it's much more practical to have the pumper talk in real absolute terms. i.e I want you to flush now because I have more context then you do.

We can then move the time delta checker in loki back up to the pumper where I think it should live and we also reduce the interface size of log back end.

barrettj12 replied:
ok, we should talk about the best way to divide the concerns here

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+// Check if it is time to send a request to remote.
+func (c *lokiClient) MaybeFlush() {
```
Not a fan of this function and what it is trying to do. It also lacks context from the caller. See comments below for further thoughts.

tlm commented on internal/overlord/logstate/pumper.go:
```diff
@@ -0,0 +1,115 @@
+		if p.parser.Next() {
+			err := p.client.WriteEntry(p.parser.Entry())
+			if err != nil {
+				logger.Noticef("couldn't write log: %v", err)
```
Needs more context and %w

tlm commented on internal/overlord/logstate/pumper.go:
```diff
@@ -0,0 +1,115 @@
+			}
+		} else {
+			if err := p.parser.Err(); err != nil {
+				logger.Noticef("couldn't parse log: %v", err)
```
Needs more context and %w

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,162 @@
+
+// send currently written request to the remote Loki
+// and reset state
+func (c *lokiClient) Flush() {
```
After reading a bit more of this PR I am not sure I am sold on the idea that the individual log backends should be logging in pebble.

I think it's much more sensible for them to be returning errors on error conditions.

They're essentially components being controlled by a manager and the manager has no visibility over if they're working and no way to perhaps do more about the situation.

barrettj12 replied:
Fair. Let's work out the best way to separate the different responsibilities of log forwarding, then hopefully we can see an obvious error flow on top of that. It's a bit hard cause multiple goroutines but we could have an error channel

barrettj12 replied:
Log clients now return errors upstream.

tlm reviewed (commented):
Thanks @barrettj12 starting to take shape. I did a partial review as I think there is probably just enough in there a lone to talk about and work on.

I still believe we should separate the buffering from the Loki implementation. In my opinion it makes for a cleaner implementation that promotes reuse.

barrettj12 commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,162 @@
+
+func parseURL(rawURL string) (*url.URL, error) {
+	u, err := url.Parse(rawURL)
+	if err != nil || u.Host == "" {
```
copied from the original PR, happy to remove

barrettj12 commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,162 @@
+	deadline *time.Time
+}
+
+func newLokiClient(service, url string, maxMessages int, delay time.Duration) (*lokiClient, error) {
```
At some point we have to parse the URL from a string into a URL object. I'm not sure where the best place to do that is

benhoyt replied:
I disagree that things like `maxMessages` should be unsigned. Go stdlib API's tend to use just `int`, and the reason is because otherwise 0 -- a very common value -- is a very problematic wraparound point. If the caller accidentally provides `x-1` and `x` is right on the boundary (i.e., 0) then you'll get a huge negative number, which is almost certainly not what you want. Whereas with `int` the wraparound occurs at -9223372036854775808 or +9223372036854775807, which is miles away from 0. Ian Lance Taylor [explains](https://groups.google.com/g/golang-nuts/c/F181gqrN0sQ/m/hQp3gQpqCQAJ) it better than I can:

> > From what I've read in the past, unsigned values are generally not
> > used because of the risk of accidental overflow bugs.
> 
> Yes, that is the reason.
> 
> Computer integers inevitably have surprising behavior for certain
> values. For signed integers those values are very large and very
> small. For unsigned integers, that value is zero. Zero is far more
> likely to occur in ordinary use than very large or very small numbers.
> If you are doing any kind of arithmetic, it's better to use signed
> integer types. You are less likely to have surprising results. 

Regarding using the `URL` type, I kinda disagree there too. That tends to be used when you're parsing/splitting URLs, not for passing in an API URL. For example, the stdlib `Get` and `Post` methods take `url string`, not a URL object. It makes them simpler to use and work with.

tlm replied:
Yep I can agree on the unsigned part. Old C++ habits die hard. 

As for the URL my reasoning is this. If we were dealing with the raw value and not doing any operations on it a string would be fine. But we are doing work on the value to add paths to get the final destination.

However happy to go either way and would rather convention that is inline with the rest of world.

barrettj12 replied:
if we're not adding defaults we can use a plain string for URL

barrettj12 commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,162 @@
+	// check for and set loki defaults if omitted from address
+	if u.Scheme == "" {
+		u.Scheme = "http"
+		if u.Port() == "" {
```
maybe. I think it's nice to put in sensible defaults cause I'm lazy. If it's gonna cause more problems than it's worth though, we can just require a fully qualified URL

barrettj12 replied:
Do you think we should disallow this?
```
log-targets:
  loki:
    location: 10.1.77.237
```
To me it seems reasonable to fill in defaults, so that the actual send URL becomes `http://10.1.77.237:3100`

tlm replied:
In this sense yes. As what we are asking for is a location and 101.77.237 doesn't describe a full location. Sure we can try and get smart and help the user out but are we helping them by adding http and 3100 when that is not the location and we just end up throwing errors further down in the code and making the user scratch their head even more.

In saying that I can see it from both sides of the conversation and I am happy to go with the flow. If I think of myself as a user I would rather just be explicit and supply the information up front to the program. I am only going to type it once if that. Saving 12 keystrokes is not the end of the world for myself v trawling through log messages to figure out why something has broken because things are being inferred on my behalf.

barrettj12 replied:
Talked to Ben and we'll just require the full location. It's easier to be strict and relax this later rather than go the other way

barrettj12 commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,162 @@
+func (c *lokiClient) Flush() {
+	c.buffer.WriteString(requestEnd)
+	//logger.Noticef(c.buffer.String())
+	resp, err := http.Post(c.lokiURL.String()+lokiPushAPI, "application/json", c.buffer)
```
kinda wanna keep the base URL separate as that gives us an option to call other API endpoints if we need to

barrettj12 replied:
`url.JoinPath` is added in Go 1.19, Pebble is still in Go 1.14 :/

barrettj12 replied:
We've decided that the user instead needs to provide the fully qualified API endpoint **including** `/loki/api/v1/push` (as Promtail does). Hence, we don't have to make assumptions about the user's Loki setup.

barrettj12 commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,162 @@
+	//logger.Noticef(c.buffer.String())
+	resp, err := http.Post(c.lokiURL.String()+lokiPushAPI, "application/json", c.buffer)
+	if err != nil {
+		logger.Noticef("error sending HTTP request: %v", err)
```
`%w` only makes sense for `fmt.Errorf`

barrettj12 replied:
fair, I can add more context to the message

benhoyt replied:
Yep, `%w` isn't meaningful for logs.

Unrelated nit: we've settled on the word "cannot" rather than "error" for these things in Pebble. Also, we capitalize log messages (but not error messages, as they may be in a "colon chain"). So this would be something like:

```go
logger.Noticef("Cannot send log to Loki: %v", err)
```

barrettj12 commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,162 @@
+	if resp.StatusCode >= 400 {
+		// TODO: don't want to reset in this case, probably want to retry
+		// but we might want to allow more logs to be written in the meantime
+		logger.Noticef("response from Loki: %v", resp.Status)
```
`resp.Status` is already a nice string containing the response code and meaning e.g.`401 Unauthorized`. `resp.StatusCode` is just the integer code. I agree more context should be added

benhoyt commented on internal/plan/plan_test.go:
```diff
@@ -24,6 +24,7 @@ import (
 	"strings"
 	"time"
 
+	"github.com/canonical/pebble/internal/testutil"
```
Nit: this should go in the canonical/pebble import group below.

benhoyt commented on internal/plan/plan.go:
```diff
@@ -597,6 +677,48 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
+
+	// Validate service log targets & apply target selections
+	for serviceName, service := range combined.Services {
+		newTargets := []string{}
```
Related: is this what's causing the changes in the tests below where you have `[]string{}` instead of leaving it nil?

benhoyt commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,165 @@
+// Register the service (and buffer) with the log manager, and create new
+// pumpers that are on the manager's to-do list.
+func (m *logManager) NewService(serviceName string, buffer *servicelog.RingBuffer) {
+	fmt.Println("new service started")
```
I usually put a TODO comment on this kind of debugging, so it's easier to spot in case I accidentally leave it in.

benhoyt commented on internal/plan/plan_test.go:
```diff
@@ -1017,3 +1043,91 @@ func (s *S) TestMarshalLayer(c *C) {
+  svc2:
+    command: cmd
+    override: merge
+    log-targets:
```
Is null the intention here, or do we want `log-targets: []`?

benhoyt commented on internal/plan/plan.go:
```diff
@@ -597,6 +677,48 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
+				}
+			}
+		}
+		service.LogTargets = newTargets
```
That said, the *logic* for creating the final selection could live in `plan.go` if we want, just in a separate function, not inside `CombineLayers`.

benhoyt commented on internal/plan/plan_test.go:
```diff
@@ -1017,3 +1043,91 @@ func (s *S) TestMarshalLayer(c *C) {
+	layer, err = plan.CombineLayers(layer)
+	c.Assert(err, IsNil)
+	for serviceName, service := range layer.Services {
+		c.Check(service.LogTargets, testutil.SameContents, expectedTargets[serviceName])
```
Instead of writing an entirely new checker, why don't we just make the code that finalizes the log targets call `sort.Strings()` on it? (Or we could sort it here, but I think stable ordering is good in any case.)

barrettj12 commented on internal/plan/plan.go:
```diff
@@ -597,6 +677,48 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
 	}
 
+	for name, target := range combined.LogTargets {
+		if target.Type != LokiType && target.Type != SyslogType {
```
i like it

hpidcock commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,303 @@
+	// mu helps manage access to network connection-related internal state.  This should be
+	// read-locked whenever using the syslog server connection and write-locked whenever
+	// replacing/reconnecting/etc the syslog server connection.
+	mu            sync.RWMutex
```
we don't need this as one client should be owned by the one forwarder

hpidcock commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,303 @@
+		client:      client,
+	}, nil
+}
+
```
Should be something like:

```
 func (f *syslogForwarder) Forward() {
 	for f.iterator.Next(f.stopCh) {
 		for f.parser.Next() {
 			f.buffer.WriteEntry(f.parser.Entry())
 		}
 		if !f.buffer.IsEmpty() {
 			f.flushBuffer()
 		}
 	}
 }
```

hpidcock replied:
iterator's Next needs to be called

barrettj12 replied:
Sure, I just copied how it was done in `internal/daemon/api_logs.go`

barrettj12 replied:
> ```
>  func (f *syslogForwarder) Forward() {
>  	for f.iterator.Next(f.stopCh) {
>  		for f.parser.Next() {
>  			f.buffer.WriteEntry(f.parser.Entry())
>  		}
>  		if !f.buffer.IsEmpty() {
>  			f.flushBuffer()
>  		}
>  	}
>  }
> ```

@hpidcock this doesn't work as written. `f.iterator.Next(f.stopCh)` never returns, so none of the log entries ever get written.

I can't see what was wrong with my way, it's done the same way as the logs API, and it was successfully able to flush logs through to Loki/syslog.

hpidcock commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+func (f *lokiForwarder) Forward() {
+	for {
```
See comments on syslog forwarder.

hpidcock commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,204 @@
+
+// Prepares a JSON request for sending to Loki.
+func (b *lokiBuffer) Request() io.Reader {
+	labelStrings := []string{}
```
This could be optimised in the future.

barrettj12 replied:
In what way?

benhoyt replied:
All the `fmt.Sprintf` and `strings.Join` and appends and whatnot are doing allocations and string formatting, so we want to avoid that as much as possible when it's a fairly tight loop (once per log, or once per flush).

This would be done by instantiating a `bytes.Buffer` once per `lokiBuffer` (but reset on flush), writing directly to that buffer, which would avoid string allocations. Also, `fmt.Sprintf` is quite slow as it has to parse the format string, allocate, and so on, so let's avoid that.

Pseudo-code:

```go
var buf bytes.Buffer  # initialize this once

# write this prefix (including labels) at the start, and after each flush
buf.Write([]byte(`{"streams":[{"stream":{"service":"foo"},"values":[`))

# write the ["timestamp","message"] part each log line
buf.Write(jsonEncodedLog)

# write this suffix before each flush
buf.Write([]byte(`]}]}`))

# to actually write to request, pass &buf directly to client.Post
```

barrettj12 replied:
ok have done this, I've used `buf.WriteString(str)` instead of `buf.Write([]byte(str))`, not sure if it makes a difference.

hpidcock commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,204 @@
+		return err
+	}
+	// Check if the response has error code
+	if resp.StatusCode >= 400 {
```
seems like if we get anything other than 200 it should be an error right? otherwise we should handle every "success" code.

barrettj12 replied:
204 is the usual response from Loki. only 4xx and 5xx reflect an error

hpidcock commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,204 @@
+		// TODO: retry?
+		b, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			b = []byte(fmt.Sprintf("(couldn't read response: %v)", err))
```
seems like a strange way to handle an error. Just return an error here.

barrettj12 replied:
At this point the request has failed, so we are already going to return an error. I am just trying to read the HTTP response to see if we can get any extra context for the error. If the read fails, just return the HTTP error, but note that we couldn't read the HTTP response to get any additional context.

benhoyt replied:
Yeah, though we might want to use a `io.LimitReader` to limit the max amount we read to 1KB or something.

hpidcock commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,204 @@
+}
+
+func (c *lokiClient) Send(body io.Reader) error {
+	resp, err := c.client.Post(c.url+lokiPushAPI, "application/json", body)
```
needs a `defer resp.Body.Close()`

hpidcock commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,217 @@
+
+// Register the service (and buffer) with the log manager, and create new
+// forwarders that are on the manager's to-do list.
+func (m *logManager) NewService(serviceName string, buffer *servicelog.RingBuffer) {
```
NewService isn't quite the right name here (as it doesn't return anything, its a notification). Something like `ForwardLogs(serviceName string, buffer *servicelog.RingBuffer, stop <-chan struct{})`, then you can also remove the StopService explicit call.

barrettj12 replied:
This is the method called by the service manager once the service has started. @benhoyt suggested ServiceStarted as a better name, does that work for you @hpidcock ?

benhoyt replied:
Yeah, I found `ForwardLogs` confusing because it sounds like a function that's going to actually do some log forwarding while you wait ... whereas it's actually "notify me that the manager has started a service", hence `ServiceStarted`. Note that it's similar in naming scheme to our other "notify" function, `PlanChanged` -- it tells you what event occurred.

hpidcock commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,217 @@
+
+		} else {
+			// Service still not started - to do later
+			stillToAdd.Add(st)
```
not sure if this is the best way to manage forwarders.

barrettj12 replied:
Ok, what's a better way? This is the awkward time **after* the plan has specified that service X should have logs forwarded to target Y, but **before** the daemon has actually started service X. We can't start log forwarding since the service ringbuffer hasn't been created yet. So we make a note to start forwarding later, once we actually receive a reference to the ringbuffer.

hpidcock commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+func (m *logManager) PlanChanged(p *plan.Plan) {
+	m.selectTargets(p)
```
I would prefer the naive approach here, if the plan changes, stop all the forwarders (let them flush), start all the forwarders again (using the same iterator so they can resume from where they left off)

barrettj12 replied:
not sure I like this, I'd rather keep them running if nothing changes. @benhoyt what do you think?

hpidcock commented on internal/overlord/servstate/handlers.go:
```diff
@@ -421,6 +421,9 @@ func (s *serviceData) startInternal() error {
 	}
 
+	// Pass buffer reference to logMgr to start log forwarding
+	s.manager.logMgr.NewService(serviceName, s.logs)
```
`s.manager.logMgr.ForwardLogs(serviceName, s.logs, done)`

hpidcock commented on internal/overlord/servstate/handlers.go:
```diff
@@ -609,6 +612,11 @@ func (s *serviceData) stop() error {
+
+	// Tell log manager to stop log forwarding for this service
+	// TODO: should this be in exited() instead?
+	s.manager.logMgr.StopService(s.config.Name)
```
don't need this, we just need to pass the stop channel in the initial call to logMgr

hpidcock reviewed (commented):
Great work, just some feedback so far

barrettj12 commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,303 @@
+func (b *syslogBuffer) WriteEntry(entry servicelog.Entry) {
+	// format defined by RFC 5424
+	timestamp := entry.Time.Format(time.RFC3339)
+	msg := fmt.Sprintf("<%d>%d %s %s %s %s %s %s %s",
```
In what way?

barrettj12 replied:
To do in future work.

benhoyt commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,206 @@
+
+	labels := map[string]string{
+		"service": serviceName,
+		"manager": "pebble",
```
Are these field names kind of standardized? How do we know these built-in labels are the right choice?

benhoyt replied:
Are we going to support custom labels defined in the plan, or is that for a subsequent iteration?

barrettj12 replied:
Seemed like a sensible default for now, I will check with Observability team. Custom labels will be supported in a future iteration.

benhoyt commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,206 @@
+	timestamp := fmt.Sprint(entry.Time.UnixNano())
+	message := strings.Trim(entry.Message, "\n")
+	encodedMsg := fmt.Sprintf("[%q,%q]", timestamp, message)
+	b.messages = append(b.messages, encodedMsg)
```
Two things:

1) This is actually JSON, and there may be subtle differences between Go's `%q` representation (which is Go syntax) and JSON string representation (which is JSON/JavaScript). In particular, Go supports `\Unnnnnnnn` Unicode escapes, whereas JSON only supports `\unnnn` ones (and would encode things above 0xFFFF as a two-character unicode combination). I realize Unicode smileys in logs is unlikely, but probably will happen in these days of Emoji everywhere.
2) This is probably quite inefficient (and done for every log line): use of `fmt.Sprint` when we know what type it is, string allocation, use of `fmt.Sprintf`, and so on.

We should probably start by fixing the correctness thing, by using `encoding/json` for at least encoding the message. That will probably slow things down a bunch, but could optimize later if needed. Correctness is first. Hopefully we can avoid the `fmt.Sprint` and `fmt.Sprintf` and that will make up for the use of json.

For a bit more efficiency, could use a `json.Encoder` attached to a `bytes.Buffer` which we reset on flush.

benhoyt commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,206 @@
+
+// Reset the buffer.
+func (b *lokiBuffer) Empty() {
+	b.messages = make([]string, 0, b.maxMessages)
```
You can reset a slice without re-allocating by doing something like this:

```go
b.messages = b.messages[:0]
```

barrettj12 replied:
Yes, or `bytes.Buffer` has a useful `Reset` method.

benhoyt commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,206 @@
+
+func newLokiClient(location string) *lokiClient {
+	return &lokiClient{
+		client: &http.Client{},
```
Should we initialize this with a `Timeout`?

barrettj12 replied:
Sure, is 10 seconds reasonable?

benhoyt commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,279 @@
+		return "-"
+	}
+	var buf bytes.Buffer
+	fmt.Fprintf(&buf, "[%s@%d", name, enterpriseNum)
```
I was going to comment on efficiency, but given this is only done once, this is fine -- thanks!

benhoyt commented on internal/plan/plan.go:
```diff
@@ -597,6 +677,43 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
+		switch target.Selection {
+		case "":
+			// Default (unspecified) selection is opt-out
+			target.Selection = OptOutSelection
```
I realize we've done this on a couple of other fields, but I think we should avoid setting values here in `CombineLayers`, otherwise a plan you submit will come back differently than what you submitted (with the default fields filled in). We want to echo back what they submitted. So I think the check for empty string / unset should occur where it's used in the log manager.

barrettj12 commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,217 @@
+				logger.Noticef("Cannot forward logs: %v", err)
+			} else {
+				m.forwarders[st] = f
+				logger.Debugf("start pumping %s -> %s/%s",
```
yep

barrettj12 commented on internal/plan/plan.go:
```diff
@@ -400,6 +408,53 @@ func (c *ExecCheck) Merge(other *ExecCheck) {
+	if other.Selection != "" {
+		t.Selection = other.Selection
+	}
+	if other.Type != "" {
```
whoopsie

barrettj12 commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,217 @@
+		for _, targetName := range service.LogTargets {
+			target, ok := p.LogTargets[targetName]
+			if !ok {
+				logger.Noticef("Log manager cannot find target %q", targetName)
```
don't think we should panic, at the end of the day we can just ignore it without having to bring down the whole system.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+) (*lokiForwarder, error) {
+
+	labels := map[string]string{
+		"pebble_service": serviceName,
```
Can we make this label a constant so we can start grouping them and write docs for their purpose?

barrettj12 replied:
Not going to bother, since it will (probably) be removed once we add custom label support.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+
+func newLokiForwarder(
+	serviceName, targetName string, target plan.LogTarget, iterator servicelog.Iterator,
+) (*lokiForwarder, error) {
```
Do we still need error in the return ?

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+func (b *lokiBuffer) WriteEntry(entry servicelog.Entry) {
+	// Write in format
+	//    ["<unix epoch in nanoseconds>","<log line>"]
+	b.buffer.WriteString("[")
```
All these lines could be trimmed down to just an fmt style command?

barrettj12 replied:
The fmt commands are much less efficient - see Ben's review comments

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+}
+
+const (
+	lokiPushAPI = "/loki/api/v1/push"
```
Can we add the word 'path' to this variable?

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+func newLokiClient(location string) *lokiClient {
+	return &lokiClient{
+		client: &http.Client{
+			Timeout: 10 * time.Second,
```
This should probably become a configurable item at some point.

barrettj12 replied:
If it proves necessary, we can easily add that to the plan later.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+}
+
+func (c *lokiClient) Send(body io.Reader) error {
+	resp, err := c.client.Post(c.url+lokiPushAPI, "application/json", body)
```
This is dangerous as mention in previous PR comments.

What you don't know here is the contents of the URL and what this environment may be doing. What if all of my Loki's are behind a load balancer and the URL I give you has a query param in it to help control the logic in my load balancer?

We need to use the appropriate GO libs here to modify the URL path and not just append to a string hope for the best. This can be done once on construction.

barrettj12 replied:
You're still making assumptions about the user's Loki setup. Ben and I have agreed that it's safer to just ask the full API endpoint, e.g. `https://232.12.67.100:3100/loki/api/v1/push`. This is what Promtail does.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+	}
+}
+
+func (c *lokiClient) Send(body io.Reader) error {
```
I am not overly keen on the separation of concerns here. The loki client is the one that understands the loki protocol but here it's just accepting a bag of bytes that we assume matches "application/json" and fire it off.

Sure you can argue that these guys are in the same package and private so it's safe. But it's really not a well thought out design as we have no real guarantees around anything at the moment.

I would suggest that we don't try and decode our buffers into the Loki format but that instead the client takes vector of strongly typed log message and does it's own heavy lifting.

barrettj12 replied:
Fixed, now we use the following interface for clients:
```go
type logClient interface {
	Send([]servicelog.Entry) error
}
```
and each client contains the backend-specific encoding logic, whereas the forwarder is generic.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+func (c *lokiClient) Send(body io.Reader) error {
+	resp, err := c.client.Post(c.url+lokiPushAPI, "application/json", body)
+	if err != nil {
+		// TODO: annotate this error with some more context?
```
Needs to be done and TODO removed.

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+	}
+
+	// Request to Loki failed
+	// TODO: retry?
```
We should retry but maybe not here. The data should probably go back into the buffer for a retry later. However what we should be establishing here is strong error messages that we can return upstream to the caller to indicate if the error received is something that can be retried.

barrettj12 replied:
Ben and I have agreed not to retry for now. We've refactored the client/forwarders as you suggested, so it will be possible to add retry logic in future (if needed).

tlm commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,250 @@
+	// Request to Loki failed
+	// TODO: retry?
+	b := make([]byte, 1024)
+	n, err := resp.Body.Read(b)
```
We need to handle n better here and error. Error could also be EOF and their may be more data available other then what n bytes were read.

We should extract all the data out and handle all the cases of error and n properly.

See https://pkg.go.dev/io#Reader

barrettj12 replied:
fair point about EOF, I'm now using an io.LimitReader instead. We don't want to extract all the data cause we don't know how large it might be. Limit it to 1KB.

barrettj12 commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,217 @@
+// Register the service (and buffer) with the log manager, and create new
+// forwarders that are on the manager's to-do list.
+func (m *logManager) NewService(serviceName string, buffer *servicelog.RingBuffer) {
+	m.buffers[serviceName] = buffer
```
In the current iteration:
- all `LogManager` fields are read/written in `PlanChanged`.
- In `ServiceStarted`, we are writing `m.buffers`, `m.iterators` and `m.forwarders`, and reading `m.planTargets`.

`PlanChanged` is called by `ServiceManager.updatePlan`, and `ServiceStarted` is called in `serviceData.startInternal`. I'm not sure if these methods run concurrently / in separate goroutines.

If we want to add a mutex, I'm thinking it's probably fine to use a shared one for the whole log manager / all the structs.

benhoyt replied:
Yeah, these can definitely be in separate goroutines. `PlanChanged` normally comes in via an API request (each of which is on its own goroutine), and `ServiceStarted` is part of the service manager state machine. Se we need a mutex. Single mutex for the whole manager sounds fine for now.

benhoyt commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,202 @@
+		return err
+	}
+
+	if s.conn != nil {
```
Won't this always be false, because we've already checked that it's not nil at the top of the function?

benhoyt commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,202 @@
+
+func (s *syslogClient) Close() error {
+	s.closed = true
+	func() {
```
Is this immediately-execute function literal unnecessary?

benhoyt commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,202 @@
+		}
+	}()
+
+	if s.conn == nil {
```
Might be simpler to put this early-return at the top, and then only do the deadline/close stuff following that.

benhoyt commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,202 @@
+	s.closed = true
+	func() {
+		if s.conn != nil {
+			s.conn.SetDeadline(time.Now().Add(-1 * time.Second))
```
Why do we need to `SetDeadline` here? Won't the `conn.Close()` below cancel any current connections?

benhoyt commented on internal/overlord/logstate/syslog_test.go:
```diff
@@ -0,0 +1,122 @@
+
+	go func() {
+		err := srv.run()
+		c.Assert(err, IsNil)
```
I don't think we call `srv.close()`? Is there a potential race condition here, where the test might finish before `srv.run()` does and this assert is called? (And then `c` might be invalid as it has no where to report the test failure.) So I think we should do a "wait for closed channel with timeout" type of thing here as well.

benhoyt commented on internal/overlord/logstate/syslog_test.go:
```diff
@@ -0,0 +1,122 @@
+		host:           "mycontainer",
+		msgid:          "-",
+		priority:       priorityVal(FacilityUserLevelMessage, SeverityInformational),
+		structuredData: syslogStructuredData("pebble", canonicalPrivEnterpriseNum, nil),
```
Shall we add a test with a couple of structured data fields while we're here? (I realize we're not using them yet, but the code is there so we should probably have a test for it.)

benhoyt commented on internal/overlord/logstate/syslog_test.go:
```diff
@@ -0,0 +1,122 @@
+		Time:    time.Date(2023, 1, 31, 1, 23, 45, 67890, time.UTC),
+		Service: "foobar",
+		Message: "test message",
+	}}
```
IMO we should send a couple of different messages (maybe three like you did with Loki tests) to tests different timestamps/messages.

benhoyt commented on internal/overlord/logstate/syslog_test.go:
```diff
@@ -0,0 +1,122 @@
+	}
+	defer conn.Close()
+
+	sc := bufio.NewScanner(conn)
```
Nit: `scanner` is probably cleaner than `sc` for this.

benhoyt commented on internal/overlord/servstate/manager_test.go:
```diff
@@ -32,6 +32,7 @@ import (
 	"time"
 	"unsafe"
 
+	"github.com/canonical/pebble/internal/servicelog"
```
Nit: pebble imports should be in the third import group.

benhoyt commented on internal/overlord/servstate/manager_test.go:
```diff
@@ -1611,3 +1612,14 @@ func (s *S) TestStopRunningNoServices(c *C) {
+
+func (f fakeLogManager) PlanChanged(p *plan.Plan) {
+	logger.Noticef("fakeLogManager notified of PlanChanged")
+	logger.Debugf("new plan: %#v", p)
```
If we're not testing logs (and we shouldn't be), should we just make these methods no-ops? Or does this help significantly in testing?

benhoyt commented on internal/plan/plan.go:
```diff
@@ -400,6 +399,52 @@ func (c *ExecCheck) Merge(other *ExecCheck) {
+const (
+	LokiType   LogTargetType = "loki"
+	SyslogType LogTargetType = "syslog"
+	UnsetType  LogTargetType = ""
```
The name `UnsetType` is a bit ambiguous. How about:

```go
LokiTarget LogTargetType = "loki"
SyslogTarget LogTargetType = "syslog"
UnsetTarget LogTargetType = ""
```

benhoyt commented on internal/servicelog/iterator.go:
```diff
@@ -69,6 +69,10 @@ func (it *iterator) Close() error {
+// Next returns true if there is more data to read from the RingBuffer.
+// If a non-nil cancel channel is passed in, Next will wait for more data to
+// become available. Sending on this channel, or closing it, will cause Next to
+// return immediately.
```
Thanks!

benhoyt commented on internal/plan/plan_test.go:
```diff
@@ -196,7 +196,8 @@ var planTests = []planTest{{
 		},
-		Checks: map[string]*plan.Check{},
+		Checks:     map[string]*plan.Check{},
+		LogTargets: map[string]*plan.LogTarget{},
```
Can we add at least a couple of basic tests for log targets and a merging operation?

benhoyt commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,185 @@
+//     iterator, to avoid log duplication.
+//   - Existing service, new target: create the forwarder using a new iterator
+//     over the existing buffer.
+//   - New service: don't create anything yet. We have to wait until
```
Helpful comment, thanks. Do we have tests for each of these three cases?

barrettj12 replied:
Yep, `TestLogManager` covers all three cases

benhoyt commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,185 @@
+//   - New service: don't create anything yet. We have to wait until
+//     ServiceStarted is called, and we are given a reference to the service's
+//     ringbuffer.
+func (m *LogManager) PlanChanged(pl *plan.Plan) {
```
I realize this function is still relatively complicated, but I think that's inherent complexity now. The flow and structure of it is a lot nicer now, thanks!

benhoyt commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,185 @@
+	}
+
+	m.forwarders = append(m.forwarders, f)
+	logger.Debugf("start forwarding logs for service %q to target %q",
```
Nit: start -> Started

benhoyt commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,185 @@
+	m.forwarders = append(m.forwarders, f)
+	logger.Debugf("start forwarding logs for service %q to target %q",
+		serviceName, targetName)
+	go f.forward()
```
Do we want to have a "log manager shutdown" method (I seem to recall Stop or Shutdown or something is called automatically on managers), to ensure graceful shutdown when Pebble itself is stopped?

barrettj12 replied:
yes, then the `LogManager` would have implement the following interfaces in the `overlord` package:
https://github.com/canonical/pebble/blob/054e5fbebc6bb0b1b195c709baa321c6c7aa2501/internal/overlord/stateengine.go#L25-L31
https://github.com/canonical/pebble/blob/054e5fbebc6bb0b1b195c709baa321c6c7aa2501/internal/overlord/stateengine.go#L40-L47

benhoyt replied:
Just the `Stop()` one, right?

barrettj12 replied:
nah, we'll need to implement both. StateStopper is only used on top of StateManager

benhoyt commented on internal/overlord/logstate/manager_test.go:
```diff
@@ -0,0 +1,246 @@
+
+func Test(t *testing.T) { TestingT(t) }
+
+func (s *managerSuite) TestSelectTargets(c *C) {
```
Nice test!

benhoyt commented on internal/overlord/logstate/manager_test.go:
```diff
@@ -0,0 +1,246 @@
+
+	// Start 3 services
+	var wg sync.WaitGroup
+	wg.Add(4)
```
It's weird that the comment says 3 but the Add says 4. I can see why below, but it might be good to adjust the comment to say why.

benhoyt commented on internal/overlord/logstate/manager_test.go:
```diff
@@ -0,0 +1,246 @@
+		})
+	}()
+	// Ensure PlanChanged gets the lock first
+	time.Sleep(1 * time.Millisecond)
```
This `time.Sleep` looks like a race condition waiting to happen. We should coordinate this. In this case can't we just call PlanChanged synchronously, without a goroutine? Or else just `Add(1)` and then `wg.Wait()` for that one first?

barrettj12 replied:
yeah I wanted to check that ServiceStarted could be called concurrently with PlanChanged, to test the lock worked. Guess it's not really necessary though.

benhoyt commented on internal/overlord/logstate/manager_test.go:
```diff
@@ -0,0 +1,246 @@
+		})
+	}()
+	// Ensure PlanChanged gets the lock first
+	time.Sleep(1 * time.Millisecond)
```
Again, we need to synchronize this rather than `time.Sleep` to avoid racy tests.

benhoyt reviewed (changes requested):
This is a big body of work and looking close now. The structure is really nice. I've added a bunch of comments. Many are minor/nits, but a few requesting small changes or adding tests -- nothing particularly structural.

Also, have you tested / will you test it against a real Loki and real syslog server and hammer it a bit?

barrettj12 commented on internal/overlord/logstate/forwarder.go:
```diff
@@ -0,0 +1,189 @@
+
+// flushBuffer sends the buffered messages to the client, then resets the buffer.
+func (f *logForwarder) flushBuffer() {
+	err := f.client.Send(f.messages)
```
I can add that, I think it's best to be defensive and do both checks though. The log client can be used independently and it doesn't make sense to send 0 messages on it.

barrettj12 replied:
Also the deadline should only be set if we have at least one message to send.

barrettj12 commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,127 @@
+		return fmt.Errorf(`
+response from Loki: %s
+%s
+`[1:], resp.Status, b)
```
Hmm, it's not really "cannot send logs to Loki", it's more like "Loki responded with an error". We already did send logs to Loki and it didn't accept them, for whatever reason.

barrettj12 commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,202 @@
+
+	// Octet framing as per RFC 5425.
+	framed := fmt.Sprintf("%d %s", len(msg), msg)
+	w.Write([]byte(framed))
```
no point, bytes.Buffer.Write never returns an error

barrettj12 commented on internal/overlord/logstate/loki_test.go:
```diff
@@ -0,0 +1,116 @@
+		"this is an even later log entry",
+	}
+
+	messages := make(chan string, 3)
```
I'll stick with the channel. The fake Loki server is also used in the manager tests, where the sending is asynchronous.

barrettj12 commented:
> Also, have you tested / will you test it against a real Loki and real syslog server and hammer it a bit?

Did some load testing today. I deployed Loki and Grafana on local microk8s using Juju. I used Pebble to run 2 services, each spitting out 1000 logs/sec. I tested the following forwarding scenarios:
- Forward logs directly to Loki.
- Run Promtail locally, receive logs from Pebble via syslog, and forward to Loki.

Both scenarios gave similar results. Pebble used about 4-10% of my CPU doing this (on my pretty-average laptop). All the logs seemed to get through to Loki.

barrettj12 commented:
There are intermittent issues with two of the tests (`forwarderSuite.TestForwarder` and `managerSuite.TestNoLogDuplication`). Both of these set up a fake backend, and expect logs to come through grouped in certain requests. Sometimes the request is missing the last log though. I guess because the forwarder fired a request before it could pull the last log. IRL this would be fine, as the last log would be sent in the next request. However these tests are failing cause they are expecting an exact request.

A couple ways to resolve this:
- "Loosen" the tests so that they don't care about how the logs are grouped - just that all the logs come through eventually.
- Play with the timeouts so we can guarantee the logs to be grouped a certain way.

benhoyt commented on internal/overlord/logstate/forwarder.go:
```diff
@@ -0,0 +1,232 @@
+// to the remote log target.
+// This method will block until the service is stopped, or stop() is called.
+// Hence, it should be run in a separate goroutine.
+func (f *logForwarder) forward() {
```
We've discussed possibly ways to simplify this via video. [Pastebin for the record.](https://pastebin.canonical.com/p/QQ5xHrw6yg/)

benhoyt commented on internal/overlord/logstate/forwarder_test.go:
```diff
@@ -0,0 +1,104 @@
+
+	writeLog("log line #5")
+	writeLog("log line #6")
+	expectLogs([]string{"log line #5", "log line #6"})
```
Are we going to be testing both the timeout and the buffer full case here? It seems to me we'll just be testing the 10ms timeout and not the buffer full case, because `expectLogs` is waiting. If so, would be good to test both cases.

barrettj12 replied:
I'll write a separate test for the buffer full case.

benhoyt commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,207 @@
+}
+
+// Send messages to remote syslog server.
+func (c *syslogClient) Send(entries []servicelog.Entry) error {
```
If we're doing the `if len(entries) == 0 { return nil }` check in Loki, we should do it here as well for consistency.

benhoyt commented on internal/overlord/logstate/loki.go:
```diff
@@ -0,0 +1,124 @@
+			// Add default labels
+			// TODO: allow specifying custom labels
+			Labels: map[string]string{
+				"pebble_service": entries[0].Service, // all entries should be from same service
```
I still think we should probably drop this from this PR, because otherwise there's this in-between state with one label, and this might get into Juju, and then people will start relying on this `pebble_service` label (which may end up with a different name when we add labels properly in the follow-up). Thoughts?

barrettj12 replied:
We need to keep it for now. This is a fundamental piece of information which will be lost if we don't have the label. IMO this is needed to make Loki support usable for now.

I would even support having `pebble_service` as a mandatory label for Loki, but this needs to be discussed further.

benhoyt replied:
Okay, I guess it's not a big deal; I just didn't want folks to start using this, set up monitoring/filtering based on this tag, and then we change it when we add proper labelling. But let's leave it for now.

benhoyt commented on internal/overlord/logstate/manager.go:
```diff
@@ -0,0 +1,203 @@
+
+// Stop implements overlord.StateStopper and stops all log forwarding.
+func (m *LogManager) Stop() {
+	wg := sync.WaitGroup{}
```
Nice idea to call all the `f.stop()`s concurrently!

barrettj12 replied:
I was thinking of doing something similar when I stop and restart them in `planChanged`. Now that I think of that, it might even be worth having a tomb for each forwarder.

benhoyt commented on internal/overlord/logstate/manager_test.go:
```diff
@@ -0,0 +1,238 @@
+	go func() {
+		defer wg.Done()
+		m.ServiceStarted("svc3", &rb)
+	}()
```
Can't we just do the following inline / synchronously, because `ServiceStarted` just creates an object an fires up the `forward` goroutine, right, so it'll be super-quick?

```
m.ServiceStarted("svc1", &rb)
m.ServiceStarted("svc2", &rb)
m.ServiceStarted("svc3", &rb)
```

barrettj12 replied:
I was trying to make this "realistic" in that Pebble can start services concurrently (I assume?). Happy to make it synchronous if you think this doesn't matter for the test.

benhoyt replied:
True, not a bad idea (and `-race` *might* pick up any issues if we're lucky). In that case maybe just a comment saying why you're doing it this way.

benhoyt reviewed (commented):
Re-reviewed. A few nits (plus the log forwarder simplification we discussed).

barrettj12 commented on README.md:
```diff
@@ -356,6 +356,53 @@ $ pebble run --verbose
+    syslog-example:
+        override: merge
+        type: syslog
+        location: tcp://0.0.0.0:1514
```
This was the actual IP address I used when I set up a syslog receiver on my local machine using rsyslog / Promtail.

barrettj12 commented on README.md:
```diff
@@ -356,6 +356,53 @@ $ pebble run --verbose
+
+By default, all services' logs will be forwarded to these targets. There are two ways to control which services' logs are sent to which targets:
+- For each log target, you can optionally specify the `selection` key. The possible values for this key are:
+  - `opt-out`: each service which doesn't **explicitly specify** log targets will have its logs forwarded to this target. This is the default behaviour.
```
"only services which explicitly specify ..." has the opposite meaning.

The full truth is, `opt-out` collects logs from all services which:
- Explicitly specify the target in `log-targets`.
- Don't specify *any* services in `log-targets`.

The logic is admittedly quite complicated here - see https://github.com/canonical/pebble/blob/f428aa12fca51136eee32774268786fbb26f77bf/internal/overlord/logstate/manager.go#L119
and while I think it's well-designed, it's difficult to explain in a clear way.

benhoyt reviewed (approved):
This is looking really good now, thanks @barrettj12! Conversation: 244. Wonder what the GitHub record is? Way more than [this one](https://github.com/torvalds/linux/pull/335). :-)

niemeyer commented on internal/overlord/logstate/forwarder.go:
```diff
@@ -0,0 +1,226 @@
+		for parser.Next() {
+			// The forward loop is either selecting on this channel or draining it
+			// in the cancel case. So no need for a select here.
+			messages <- parser.Entry()
```
The machinery that decouples the iterator, the parser, the buffer inside the forward loop, and the client, seem to add signifciant complexity that we don't seem to benefit from. Let's just take a step back and consider the overall flow.

At the moment we have data buffered at those layers:

- The ring buffer
- (NOT the iterator, it's just a pointer to the ring)
- The servicelog.Parser
- The buffer inside forward()
- The logClient (whether persistent or not, there's a data buffer there while sending)

All of those intermediate layers are just processing data from one side to the other. It might be tempting to buffer data in between to make sure the system does not run out of memory, well, except both sides are _already in memory!_  So any attempt to buffer data in the middle in an intermediate state is actually _increasing_ the amount of memory used in the overall system. If we want less memory consumption, all we need to do is to consume more slowly from the left side, and when we do consume, send it all the way to the right side.

Assuming such a design, we'd need a single goroutine that parses out of the iterator, and any time it manages to parse anything it adds to the client which can then serialize to its buffer and keep it pretty much ready for dispatch. Straight all the way from ring buffer to the output buffer.

Then, there's apparently another issue in the current system: there is an independent client for each forwarder, which means if we have 10 services and 2 targets, we have 20 different connections, each with their own series of buffers. Is that the case?  If so, that really doesn't feel so great.

So, the recommended simplified design would be something like this: One single client per log target for all of Pebble, no intermediate buffers.  All the data that leaves the ring buffer is processed on one single control goroutine, and lands on the client buffer as ready to dispatch as possible, so we know the actual amount of bytes we have to send, and are not holding references to Go structures. One _Push_ blocking method to add entries into the client which also flushes when above a given size limit, and a second method _Flush_ to force it empty.  These methdos make it easy for the client to tolerate concurrency properly (any number of goroutines pushing their messages will naturally block and wait when a size limit is exceeded. To handle the time-based delay, we can have a second method like _FlushAfter(duration),_ where the timing of last flush is held internally by the client, which means concurrency would also work fine if multilpe goroutines are recommending the flush at once.

And that would be all, I think.  Of course, it's a bit hard to keep the whole picture at once in memory, but hopefully I'm not missing major points here.

niemeyer reviewed (changes requested):
Thanks for the great work here, everyone.  I have a few recommendations below to get started, but as a general overview this is looking very nice. In particular, the whole plan side of things looks pretty much ready to land, while on the forwarding implementation there might be a couple of details for us to figure out.

Also on that basis, another recommendation here is for us to break down this PR into smaller parts that can more easily evolve and get merged independently. Considering the design in place, I suggest the following parts:

1) The entire parsing and configuration aspects, also including a stub manager that does almost nothing other than reading and validating targets.
2) The manager implementation, which includes all the glue, but no real clients.
3) The syslog client.
4) The loki client.
 
Step 1 is probably most of the code, and it's also mostly ready to merge, so we have some quick wins ahead of us. Step 2 is likely where most of our conversations will take place. Step 3 and 4 should be easy when we get there.

How does that sound?

niemeyer commented on internal/overlord/logstate/syslog.go:
```diff
@@ -0,0 +1,210 @@
+
+	// Octet framing as per RFC 5425.
+	framed := fmt.Sprintf("%d %s", len(msg), msg)
+	c.data.Write([]byte(framed))
```
Please note that considering other comments it migth be worth waiting this one out until the core is sorted.

benhoyt commented on internal/plan/plan.go:
```diff
@@ -178,14 +183,7 @@ func (s *Service) Merge(other *Service) {
-		return true
-	}
-	return reflect.DeepEqual(s, other)
+	s.LogTargets = append(s.LogTargets, other.LogTargets...)
```
I think de-duping these is a reasonable thing to do, however, I suggest that `plan.go` should just be about layers and configuration and shouldn't have smarts about what the fields mean and whether specific fields need to be de-duped. So I suggest we do the de-duping but in the log manager.

niemeyer replied:
The algorithm for merging layers is entirely inside the plan. Are we talking about the same thing?

benhoyt commented on internal/overlord/logstate/forwarder.go:
```diff
@@ -0,0 +1,226 @@
+
+// stop interrupts the main forward loop, telling the forwarder to shut down.
+// Once this method returns, it is safe to re-use the iterator.
+func (f *logForwarder) stop() {
```
Sounds reasonable to use tombs. However, we should be careful with error handling; I don't think we don't want a single error to be bubbled up and end up restarting the forwarder. We did have errors "buried in logs only", but that was by design -- without retry logic that's about all you can do, and when new logs come in, we'll attempt to send those. We wanted to do the initial version without retry logic, as that adds complexity and raises a number of other questions (what if logs get backed up when the target is error-ing, how long / how many do we keep, and so on). In any case, we'll consider this when we switch to tombs.

niemeyer replied:
Yes, there's still a distinction between a fatal error a non-fatal error. Right now neither is surfacing.

barrettj12 commented on internal/overlord/servstate/manager.go:
```diff
@@ -466,6 +475,19 @@ func (m *ServiceManager) Replan() ([]string, []string, error) {
+	oldCopy.LogTargets = nil
+	newCopy := new.Copy()
+	newCopy.LogTargets = nil
+	return !reflect.DeepEqual(oldCopy, newCopy)
```
The problem we're trying to solve here is that not all changes to `plan.Service` require a service restart. In particular, when we added the `LogTargets` field to `plan.Service`, we noticed in tests that **only** changing the `LogTargets` would trigger a service restart. There's no reason this should happen, since the service **itself** has not changed - only where its logs are being sent, and this is managed independently by the log manager. (Changes to service metadata and some other fields shouldn't require a restart either, but this is not of primary concern right now).

Hence, we created a `NeedsRestart` method that checks if there has been a **fundamental** change to a service which would necessitate a restart. This is more appropriate for the use case (in `Replan`, we we using `service.Equal` to determine whether to restart a service). `NeedsRestart` seems more appropriate inside the `servstate` package, as it's concerned with the semantics of the service config from the POV of the service manager, and the `plan` package should be decoupled from the semantics. The only use of the `service.Equal` method was inside `Replan`, hence we removed the unused method.

If your concerns are more about the implementation of `NeedsRestart`, then fair enough, we can rewrite this so that we are not deep-copying the service structs.

niemeyer replied:
We're mostly in agreement, except here:

> NeedsRestart seems more appropriate inside the servstate package, as it's concerned with the semantics of the service config from the POV of the service manager, and the plan package should be decoupled from the semantics.

That doesn't seem to be the case. The analysis being done tells if a service setup has been modified in a way that changes what should be started. This is a static analysis of the plan that can be done completely by itself, with no regards for what the manager or the system state is at all. That was already the case before with `service.Equal`, and it remains the case even if we're tuning its meaning because now there are references to external objects that do not modify how the process is started.

It may look like this:
```
// SameStart returns whether s is started the same way as other.
func (s *Service) SameStart(other *Service) bool { ... }
```

barrettj12 commented:
Hi @niemeyer, thanks a lot for the review. I've addressed most of your review comments and opened the first PR #209 as you suggested, focusing just on the changes to the plan. Once this merges, I can start working on the simplified log forwarding design as you suggested.

jnsgruk commented:
@barrettj12 @hpidcock what's the status here - we could do with progressing this. Perhaps we should open a new PR to address the remaining features, and close this one to avoid any further confusion?

niemeyer commented:
PR #209 has been reviewed and merged with the first phase. I'm going to close this PR to make the status more clear, as it needs to be broken down into the remaining parts as outlined in the review above.
